cmd/remote-session.go

```diff
@@ -0,0 +1,225 @@
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+// Destroys the "remote session". In reality it just deletes
+// the container referenced by $COREOS_ASSEMBLER_REMOTE_SESSION.
+func runDestroy(c *cobra.Command, args []string) error {
+	if err := checkEnv(); err != nil {
+		return err
+	}
+	podmanargs := strings.Fields(fmt.Sprintf(
+		"--remote rm -f %s",
+		os.Getenv("COREOS_ASSEMBLER_REMOTE_SESSION"),
+		strings.Join(args, " ")))
+	cmd := exec.Command("podman", podmanargs...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+// Executes a command in the "remote session". Mostly just a
+// `podman --remote exec`.
+func runExec(c *cobra.Command, args []string) error {
+	if err := checkEnv(); err != nil {
+		return err
+	}
+	podmanargs := strings.Fields(fmt.Sprintf(
+		"--remote exec -it %s cosa %s",
+		os.Getenv("COREOS_ASSEMBLER_REMOTE_SESSION"),
+		strings.Join(args, " ")))
+	cmd := exec.Command("podman", podmanargs...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			// If the command failed let's exit with the same exitcode
+			// as the remotely executed process.
+			os.Exit(exitError.ExitCode())
+		} else {
+			return err
+		}
+	}
+	return nil
+}
+
+// Executes a `podman --remote ps -a --filter id=<container>`
+// to show the status of the remote running cosa container.
+func runPS(c *cobra.Command, args []string) error {
+	if err := checkEnv(); err != nil {
+		return err
+	}
+	podmanargs := strings.Fields(fmt.Sprintf(
+		"--remote ps -a --filter id=%s",
+		os.Getenv("COREOS_ASSEMBLER_REMOTE_SESSION")))
+	cmd := exec.Command("podman", podmanargs...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+// runSync provides an rsync-like interface that allows
+// files to be copied to/from the remote. It uses
+// `podman --remote exec` as the transport for rsync (see [1])
+//
+// One of the arguments here must be prepended with a `:`. This
+// argument will represent the path on the remote. This function
+// will substitute `:` with `$COREOS_ASSEMBLER_REMOTE_SESSION:`.
+// That environment var just contains the container ID on the remote.
+//
+// [1] https://github.com/moby/moby/issues/13660
+func runSync(c *cobra.Command, args []string) error {
+	// check environment
+	if err := checkEnv(); err != nil {
+		return err
+	}
+	// check arguments. Need one with pre-pended ':'
+	found := 0
+	for index, arg := range args {
+		if strings.HasPrefix(arg, ":") {
+			args[index] = fmt.Sprintf("%s%s",
+				os.Getenv("COREOS_ASSEMBLER_REMOTE_SESSION"), arg)
+			found++
+		}
+	}
+	if found != 1 {
+		return fmt.Errorf("Must pass in a single arg to rsync with `:` prepended")
+	}
+	// build command and execute
+	rsyncargs := strings.Fields(fmt.Sprintf(
+		"-avh --mkpath --blocking-io %s --rsh", strings.Join(args, " ")))
```

Comment:
Hmmm...some potential whitespace/shell quoting issues here no?  One thing I wanted to use in bashexec is https://pkg.go.dev/github.com/frioux/shellquote but that requires more duplicate vendoring...

(This issue applies in the other places we're "parsing" shell via `strings.Fields`, see also https://github.com/coreos/coreos-assembler/pull/2959#discussion_r913784500 )

Reply:
Definitely; looks much better to me now!

By way of example: here's a setroubleshoot CVE I remember going by: https://bugzilla.redhat.com/show_bug.cgi?id=1332644
:arrow_right:  https://github.com/fedora-selinux/setroubleshoot/commit/5cd60033ea7f5bdf8c19c27b23ea2d773d9b09f5

This one was embarrassing because it's part of the security system *introducing* a flaw.  And while this code isn't security sensitive, you never know when the next person will copy/paste or reuse a library function in a place that is.

Reply:
Also, command line injection is one of those flaws that's still easy to do in memory safe languages, so we need to be on alert for it and not just think that e.g. because the service is in Go/Rust we're safe.